File: frontend/src/api/canvas.ts
/** 画布(生图工作台)API:文档 CRUD + 回滚 + 分镜导入 + 批量生成。 */
import { api } from './client'
import type { CanvasEdge, CanvasNode } from '../lib/canvasTypes'

export interface CanvasBoardDocument {
  nodes: CanvasNode[]
  edges: CanvasEdge[]
}

export interface CanvasBoardDto {
  id: string
  project_id: string | null
  name: string
  document: CanvasBoardDocument
  version: number
  created_at: string
  updated_at: string
}

export interface CanvasNodeGenerateOut {
  node_id: string
  task_id: string | null
  keyframe_id: string | null
  error: string | null
}

export interface CanvasBoardGenerateOut {
  task_id: string
  nodes: CanvasNodeGenerateOut[]
}

export interface ProjectLite {
  id: string
  title: string
}

export interface SegmentLite {
  id: string
  episode_id: string
  index: number
  description: string | null
  shot_type: string | null
}

export function listBoards(): Promise<CanvasBoardDto[]> {
  return api.get('/canvas')
}

export function createBoard(body: { name?: string; project_id?: string | null }): Promise<CanvasBoardDto> {
  return api.post('/canvas', body)
}

export function getBoard(id: string): Promise<CanvasBoardDto> {
  return api.get(`/canvas/${id}`)
}

export function saveBoard(id: string, body: { document: CanvasBoardDocument; name?: string }): Promise<CanvasBoardDto> {
  return api.put(`/canvas/${id}`, body)
}

export function rollbackBoard(id: string): Promise<CanvasBoardDto> {
  return api.post(`/canvas/${id}/rollback`)
}

export function deleteBoard(id: string): Promise<void> {
  return api.del(`/canvas/${id}`)
}

export function importBoardFromSegment(segmentId: string, name?: string): Promise<CanvasBoardDto> {
  return api.post('/canvas/import-segment', { segment_id: segmentId, name })
}

export function importBoardsFromSegments(segmentIds: string[], name?: string): Promise<CanvasBoardDto> {
  return api.post('/canvas/import-segments', { segment_ids: segmentIds, name })
}

export function updateSegmentCompositionPoint(segmentId: string, point: string): Promise<unknown> {
  return api.put(`/segments/${segmentId}`, { composition_point: point })
}

export function generateBoard(
  id: string,
  body: { node_ids: string[]; model_id?: string; ratio?: string; kind?: 'image' | 'video' },
): Promise<CanvasBoardGenerateOut> {
  return api.post(`/canvas/${id}/generate`, body)
}

// ---- 导演台模式（多段连续生视频,2026-08-29）----
export interface DirectorShotIn {
  node_id: string
  prompt?: string
  duration_sec?: number
  from_prev?: boolean
}

export interface DirectorGenerateBody {
  node_ids: string[]
  config?: {
    task_type?: string
    ratio?: string
    res?: string
    fps?: number
    steps?: number
    sampler?: string
    scheduler?: string
    shift_video?: number
    shift_audio?: number
    cfg?: number
    seed?: number
    context_enabled?: boolean
    context_frames?: number
    global_prompt?: string
    model_id?: string
    /** ComfyUI 式方案节点自身 id（2026-08-30）：后端把产物回写到该节点做节点内预览 */
    scheme_node_id?: string
  }
  shots?: DirectorShotIn[]
}

export interface DirectorGenerateOut {
  task_id: string
  board_id: string
  node_ids: string[]
  message?: string
}

export function directorGenerate(
  id: string,
  body: DirectorGenerateBody,
): Promise<DirectorGenerateOut> {
  return api.post(`/canvas/${id}/director-generate`, body)
}

// ---- 画布生成方案体系（2026-08-29）：选择方案 → 统一执行 ----
export interface SchemeMeta {
  key: string
  label: string
  description: string
  input_kind: 'node-batch' | 'director-segments' | string
}

export interface SchemeGenerateOut {
  task_id: string
  board_id: string
  scheme: string
  node_ids: string[]
  nodes: Array<Record<string, unknown>>
  message?: string
}

export function listCanvasSchemes(): Promise<SchemeMeta[]> {
  return api.get('/canvas/schemes')
}

export function schemeGenerate(
  id: string,
  body: {
    scheme: string
    node_ids?: string[]
    kind?: 'image' | 'video'
    model_id?: string
    ratio?: string
    config?: DirectorGenerateBody['config']
    shots?: DirectorGenerateBody['shots']
  },
): Promise<SchemeGenerateOut> {
  return api.post(`/canvas/${id}/scheme-generate`, body)
}

export async function findBoardForSegment(segmentId: string): Promise<string | null> {
  const boards = await listBoards()
  for (const b of boards) {
    const doc = b.document as unknown as { nodes?: Array<{ type?: string; data?: { segmentId?: string } }> }
    if (doc?.nodes?.some((n) => n.type === 'shot' && n.data?.segmentId === segmentId)) return b.id
  }
  return null
}

export interface CanvasAuditItem {
  node_id: string
  label: string
  issue: 'failed' | 'unbound' | 'no_product' | 'no_point'
  detail?: string
}

export interface CanvasAudit {
  board_id: string
  items: CanvasAuditItem[]
  running_tasks: number
  failed_tasks: number
  recent_failed: { task_id: string; error?: string }[]
}

export function auditBoard(id: string): Promise<CanvasAudit> {
  return api.get(`/canvas/${id}/audit`)
}

export function enhanceSegmentPrompt(
  segmentId: string,
  prompt?: string,
  target: 'image' | 'video' = 'image',
): Promise<{ enhanced_prompt: string; negative_prompt: string }> {
  return api.post(`/segments/${segmentId}/enhance-prompt`, { prompt: prompt ?? null, target })
}

export function listProjects(): Promise<ProjectLite[]> {
  return api.get('/projects')
}

export function listProjectSegments(projectId: string): Promise<SegmentLite[]> {
  return api.get(`/projects/${projectId}/segments`)
}

File: backend/app/schemas/canvas.py
"""画布(生图工作台)Schemas。"""
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class CanvasBoardCreate(BaseModel):
    name: str = Field(default="导演画布", max_length=120)
    project_id: UUID | None = None


class CanvasBoardSave(BaseModel):
    document: dict
    name: str | None = None


class CanvasSegmentImport(BaseModel):
    segment_id: UUID
    name: str | None = None


class CanvasSegmentsImport(BaseModel):
    """多分镜合并导入(M3 导演层):一次把多个分镜(可跨幕)落成一张画布。"""

    segment_ids: list[UUID]
    name: str | None = None


class CanvasBoardGenerate(BaseModel):
    node_ids: list[str]
    model_id: UUID | None = None
    ratio: str | None = None
    # M2:image=关键帧(默认)| video=镜头视频(R2V 直接出片,复用 video_service 链路)
    kind: str = "image"


class CanvasNodeGenerateOut(BaseModel):
    node_id: str
    task_id: UUID | None = None
    keyframe_id: UUID | None = None
    error: str | None = None


class CanvasBoardGenerateOut(BaseModel):
    task_id: UUID
    nodes: list[CanvasNodeGenerateOut]


class CanvasBoardOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    project_id: UUID | None
    name: str
    document: dict
    version: int
    created_at: datetime
    updated_at: datetime
class DirectorShot(BaseModel):
    """导演台分段脚本：一行=一段镜（顺序=时间顺序）。"""

    node_id: str
    prompt: str | None = None  # 缺省由后端 LLM 视频增强补全
    duration_sec: float | None = None  # None/<=0 → 后端按分镜时长换算
    from_prev: bool = True  # 引用上段（段间引导，需 config.context_enabled）


class DirectorGenerate(BaseModel):
    """画布导演台模式请求体（与 generate 的 node_ids 语义不同：
    这里是『一段连续影片』的镜头清单，不是批量分别生成）。
    """

    node_ids: list[str]
    config: dict | None = None  # task_type/ratio/res/fps/steps/sampler/scheduler/shift_video/shift_audio/context_enabled/context_frames/refine
    shots: list[DirectorShot] | None = None  # 缺省则按 node_ids 顺序自动生成


class DirectorGenerateOut(BaseModel):
    task_id: UUID
    board_id: UUID
    node_ids: list[str]
    message: str | None = None

# ── 画布生成方案体系（2026-08-29）──────────────────────────


class SchemeMeta(BaseModel):
    """方案元数据（GET /canvas/schemes 用）：前端渲染方案菜单。"""

    key: str
    label: str
    description: str = ""
    input_kind: str = "node-batch"  # node-batch | director-segments | ...


class SchemeGenerate(BaseModel):
    """统一方案执行请求体：scheme 分发 + 通用字段 + 方案私有 config。"""

    scheme: str
    node_ids: list[str] = Field(default_factory=list)
    kind: str = "image"  # single 方案：image=关键帧 | video=镜头视频
    model_id: UUID | None = None
    ratio: str | None = None
    config: dict | None = None  # 方案私有参数（director 的 ratio/res/context 等）
    shots: list[DirectorShot] | None = None  # director 分段脚本


class SchemeGenerateOut(BaseModel):
    task_id: UUID
    board_id: UUID
    scheme: str
    node_ids: list[str]
    nodes: list[dict] = Field(default_factory=list)  # single: 逐节点结果
    message: str | None = None


